modules/standard/Random.chpl

```diff
@@ -1208,143 +1208,425 @@ module Random {
+      // just computed the RNGs at the later incs
+      param numGenForResultType = numGenerators(resultType);
+      param numGen = states.size;
+      if numGenForResultType > numGen then
+        compilerError("PCGRandomStream cannot produce " +
+                      resultType:string +
+                      " (requiring " +
+                      (32*numGenForResultType):string +
+                      " bits) from a stream configured for " +
+                      (32*numGen):string +
+                      " bits of output");
 
-     Low-level PCG RNG implementation.
+    }
 
-     Besides a Chapel :class:`RandomStream` interface, the :mod:`PCGRandomLib`
-     module includes a number of low-level PCG random functions.
+    // Wrapper that takes a result type
+    private inline
+    proc randlc(type resultType, ref states) {
 
-       * :record:`pcg_setseq_64_xsh_rr_32_rng` (the default PCG RNG)
-       * :record:`pcg_setseq_64_rxs_m_xs_64_rng`
-       * :record:`pcg_setseq_32_rxs_m_xs_32_rng`
-       * :record:`pcg_setseq_16_rxs_m_xs_16_rng`
-       * :record:`pcg_setseq_8_rxs_m_xs_8_rng`
-       * :record:`pcg_setseq_N_rxs_m_xs_N_rng` which is a generalization of the
-         above
+      checkSufficientBits(resultType, states);
 
-     These names come from the PCG paper and reference implementations.  The
-     first integer is the number of state bits, and the last integer is the
-     number of output bits. The other parts describe the permutation function in
-     the LCG.
+      if resultType == complex(128) {
+        return (randToReal64(rand64_1(states)),
+                randToReal64(rand64_2(states))):complex(128);
+      } else if resultType == complex(64) {
+        return (randToReal32(rand32_1(states)),
+                randToReal32(rand32_2(states))):complex(64);
+      } else if resultType == imag(64) {
+        return _r2i(randToReal64(rand64_1(states)));
+      } else if resultType == imag(32) {
+        return _r2i(randToReal32(rand32_1(states)));
+      } else if resultType == real(64) {
+        return randToReal64(rand64_1(states));
+      } else if resultType == real(32) {
+        return randToReal32(rand32_1(states));
+      } else if resultType == uint(64) || resultType == int(64) {
+        return rand64_1(states):resultType;
+      } else if resultType == uint(32) || resultType == int(32) {
+        return rand32_1(states):resultType;
+      } else if(resultType == uint(16) ||
+                resultType == int(16)) {
+        return (rand32_1(states) >> 16):resultType;
+      } else if(resultType == uint(8) ||
+                resultType == int(8)) {
+        return (rand32_1(states) >> 24):resultType;
+      } else if isBoolType(resultType) {
+        return (rand32_1(states) >> 31) != 0;
+      }
+    }
 
-     */
-    module PCGRandomLib {
-    /*
-       Low-level PCG random number generation interface (64-bits of state,
-       32-bits output).
+    // returns x with min <= x <= max (for integers)
```

Comment:
I can't make sense of this diff, so am relying on it simply being a code reordering and not reviewing further.  I could imagine going further and putting PCG ahead of NPB in the file on the argument that it's the dominant case for those reading the source.